src/tensor_ops/convtrans2d/mod.rs

```diff
@@ -0,0 +1,584 @@
+        let mut out = lhs.device.try_zeros()?;
+        lhs.device.forward(op, &lhs, &rhs, &mut out)?;
+        let phantom_out = out.clone();
+        tape.try_alloc_grad(&lhs)?;
+        tape.try_alloc_grad(&rhs)?;
+        tape.try_alloc_grad(&out)?;
+        tape.add_backward_op(move |grads| {
+            let (grad_lhs, grad_rhs, grad_out) = grads.muts_and_ref(&lhs, &rhs, &phantom_out);
+            lhs.device
+                .backward(op, &lhs, grad_lhs, &rhs, grad_rhs, &phantom_out, grad_out)
+        });
+        Ok(out.put_tape(tape))
+    }
+}
+
+impl<
+        B: Dim,
+        const C: usize,
+        const H: usize,
+        const W: usize,
+        const O: usize,
+        const K: usize,
+        const S: usize,
+        const P: usize,
+        E: Dtype,
+        D: ConvTrans2DKernel<E> + ZerosTensor<E>,
+        T: 'static + Tape<E, D>,
+    > TryConvTrans2DTo<Tensor<Rank4<O, C, K, K>, E, D>, S, P>
+    for Tensor<(B, Const<C>, Const<H>, Const<W>), E, D, T>
+where
+    Const<H>: ConvTransAlgebra<K, S, P>,
+    Const<W>: ConvTransAlgebra<K, S, P>,
+{
+    type Output = Tensor<
+        (
+            B,
+            Const<O>,
+            <Const<H> as ConvTransAlgebra<K, S, P>>::Convolved,
+            <Const<W> as ConvTransAlgebra<K, S, P>>::Convolved,
+        ),
+        E,
+        D,
+        T,
+    >;
+    fn try_convtrans2d_to(
+        self,
+        filters: Tensor<Rank4<O, C, K, K>, E, D>,
+    ) -> Result<Self::Output, Self::Err> {
+        let batch = self.shape().0;
+        let op = ConvTrans2DOp::new(S, P, K, [batch.size(), C, H, W], O);
+        let (lhs, ltape) = self.split_tape();
+        let (rhs, rtape) = filters.split_tape();
+        let mut out =
+            lhs.device
+                .try_zeros_like(&(batch, Const, Default::default(), Default::default()))?;
+        let mut tape = ltape.merge(rtape);
+        lhs.device.forward(op, &lhs, &rhs, &mut out)?;
+        let phantom_out = out.clone();
+        tape.try_alloc_grad(&lhs)?;
+        tape.try_alloc_grad(&rhs)?;
+        tape.try_alloc_grad(&out)?;
+        tape.add_backward_op(move |grads| {
+            let (grad_lhs, grad_rhs, grad_out) = grads.muts_and_ref(&lhs, &rhs, &phantom_out);
+            lhs.device
+                .backward(op, &lhs, grad_lhs, &rhs, grad_rhs, &phantom_out, grad_out)?;
+            Ok(())
+        });
+        Ok(out.put_tape(tape))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{tensor::*, tensor_ops::*, tests::*};
+
+    #[test]
+    /// TODO
+    /// Produced by
+    /// ```python
+    /// x = torch.rand((2, 3, 3), requires_grad=True)
+    /// w = torch.rand((2, 3, 3, 3), requires_grad=True)
+    /// print(x)
+    /// print(torch.swapaxes(w, 0, 1))
+    /// y = torch.conv_transpose2d(x, w)
+    /// print(y)
+    /// y.exp().mean().backward()
+    /// print(x.grad)
+    /// print(torch.swapaxes(w.grad, 0, 1))
+    /// ```
+    fn convtrans2d_test() {
```

Comment:
any thoughts on more easily writing tests & increasing code coverage for complex operations like conv2d & conv transpose? They are some of the more complicated ones to test (along with transformers), and the current method we have for testing leaves a lot to be desired

Reply:
I think as long as the tests have a tensor with all dimensions of input and output >1 and both the output and grads match with pytorch, I think its fine. Were you thinking more along the lines of tests that don't rely on pytorch/ are randomly generated?

Reply:
Yeah exactly, and in a way where we can test larger tensors. Mostly thinking about testing conv2d in cuda. The kernels are fairly complex, and some issues might not be caught unless we have fairly large sizes on things.